Source/SecureChat/KeysRotation/KeysRotator.swift

```diff
@@ -0,0 +1,278 @@
+    private let longTermKeyTtl: TimeInterval
+    private let outdatedLongTermKeyTtl: TimeInterval
+    private let desiredNumberOfOneTimeKeys: Int
+    private let longTermKeysStorage: LongTermKeysStorage
+    private let oneTimeKeysStorage: OneTimeKeysStorage
+    private let client: RatchetClientProtocol
+    private let mutex = Mutex()
+    private let keyUtils = RatchetKeyUtils()
+
+    /// Initializer
+    ///
+    /// - Parameters:
+    ///   - identityPrivateKey: identity private key
+    ///   - identityCardId: identity card id
+    ///   - orphanedOneTimeKeyTtl: time that one-time key lives in the storage after been marked as orphaned. Seconds
+    ///   - longTermKeyTtl: time that long-term key is been used before rotation. Seconds
+    ///   - outdatedLongTermKeyTtl: time that long-term key lives in the storage after been marked as outdated. Seconds
+    ///   - desiredNumberOfOneTimeKeys: desired number of one-time keys
+    ///   - longTermKeysStorage: long-term keys storage
+    ///   - oneTimeKeysStorage: one-time keys storage
+    ///   - client: RatchetClient
+    public init(identityPrivateKey: VirgilPrivateKey,
+                identityCardId: String,
+                orphanedOneTimeKeyTtl: TimeInterval = 24 * 60 * 60,
+                longTermKeyTtl: TimeInterval = 5 * 24 * 60 * 60,
+                outdatedLongTermKeyTtl: TimeInterval = 24 * 60 * 60,
+                desiredNumberOfOneTimeKeys: Int = 100,
+                longTermKeysStorage: LongTermKeysStorage,
+                oneTimeKeysStorage: OneTimeKeysStorage,
+                client: RatchetClientProtocol) {
+        self.identityPrivateKey = identityPrivateKey
+        self.identityCardId = identityCardId
+        self.orphanedOneTimeKeyTtl = orphanedOneTimeKeyTtl
+        self.longTermKeyTtl = longTermKeyTtl
+        self.outdatedLongTermKeyTtl = outdatedLongTermKeyTtl
+        self.desiredNumberOfOneTimeKeys = desiredNumberOfOneTimeKeys
+        self.longTermKeysStorage = longTermKeysStorage
+        self.oneTimeKeysStorage = oneTimeKeysStorage
+        self.client = client
+    }
+
+    /// Rotates keys
+    ///
+    /// Rotation process:
+    ///         - Retrieve all one-time keys
+    ///         - Delete one-time keys that were marked as orphaned more than orphanedOneTimeKeyTtl seconds ago
+    ///         - Retrieve all long-term keys
+    ///         - Delete long-term keys that were marked as outdated more than outdatedLongTermKeyTtl seconds ago
+    ///         - Check that all relevant long-term and one-time keys are in the cloud
+    ///             (still persistent in the cloud and were not used)
+    ///         - Mark used one-time keys as used
+    ///         - Decide on long-term key roration
+    ///         - Generate needed number of one-time keys
+    ///         - Upload keys to the cloud
+    ///
+    /// - Returns: GenericOperation
+    public func rotateKeysOperation() -> GenericOperation<Void> {
```

Comment:
Function fits in 160+ lines. It should be splitted.

Reply:
In this case breaking function in parts will make it harder to understand. See here https://github.com/VirgilSecurity/virgil-sdk-pfs-x/blob/master/Source/SecureChat/KeysRotator.swift#L83